### PR TITLE
debug: Converted uint64-blocknumber into rpc.Blocknumber

### PR DESCRIPTION
**Pull request by holiman**
So now you can do 

```
> debug.traceBlockByNumber("3470000")
{
  error: "",
  structLogs: [],
  validated: true
}
```
Stil can't do 
``` 
> debug.traceBlockByNumber(3470000)
Error: invalid argument 0: invalid blocknumber 3.47e+06
    at web3.js:3104:20
    at web3.js:6191:15
    at web3.js:5004:36
```
But this also failed previously, like this: 
```
> debug.traceBlockByNumber(3470000)
Error: invalid argument 0: json: cannot unmarshal number 3.47e+06 into Go value of type uint64
    at web3.js:3104:20
    at web3.js:6191:15
    at web3.js:5004:36
    at <anonymous>:1:1
```

### CODE REVIEW

File: eth/api.go [hunk 1]
```diff
@@ -349,11 +361,22 @@ func (api *PrivateDebugAPI) TraceBlockFromFile(file string, config *vm.LogConfig
 }
 
 // TraceBlockByNumber processes the block by canonical block number.
-func (api *PrivateDebugAPI) TraceBlockByNumber(number uint64, config *vm.LogConfig) BlockTraceResult {
+func (api *PrivateDebugAPI) TraceBlockByNumber(blockNr rpc.BlockNumber, config *vm.LogConfig) BlockTraceResult {
+
+	var block *types.Block
```
Comment (karalabe) [inline eth/api.go]: Please move this below the comment, and delete the empty line.

File: eth/api.go [hunk 2]
```diff
@@ -281,10 +282,21 @@ func NewPublicDebugAPI(eth *Ethereum) *PublicDebugAPI {
 }
 
 // DumpBlock retrieves the entire state of the database at a given block.
-func (api *PublicDebugAPI) DumpBlock(number uint64) (state.Dump, error) {
-	block := api.eth.BlockChain().GetBlockByNumber(number)
+func (api *PublicDebugAPI) DumpBlock(blockNr rpc.BlockNumber) (state.Dump, error) {
+
```
Comment (karalabe) [inline eth/api.go]: Please remove this empty line.

File: eth/api.go [hunk 3]
```diff
@@ -281,10 +282,21 @@ func NewPublicDebugAPI(eth *Ethereum) *PublicDebugAPI {
...
+	var block *types.Block
+
+	switch blockNr {
+	case rpc.PendingBlockNumber:
+		// Pending block is only known by the miner
+		block = api.eth.miner.PendingBlock()
```
Comment (karalabe) [inline eth/api.go]: This code path will need a bit more special handling. The "dumping" code below in line 301 looks up the state from the real blockchain. If we're dumping the pending state, you actually need to request both the pending block as well as the pending state from the miner and operate on those versus the current approach.

Comment (karalabe) [inline eth/api.go]: `block, state := api.eth.miner.Pending()`

File: core/blockchain.go [hunk 4]
```diff
@@ -236,6 +236,10 @@ func (self *BlockChain) loadLastState() error {
 	// Try to be smart and issue a pow verification for the head to pre-generate caches
 	go self.pow.Verify(types.NewBlockWithHeader(currentHeader))
 
+	log.Info("Dumping code...")
+	//self.stateCache.CodeDump()
+	log.Info("Code dumped")
```
Comment (karalabe) [inline core/blockchain.go]: Erm? :D

**Comment by fjl** [comment]
Side note: can you please pay some attention to the commit prefixes? There is no 'debug' package. The change is in package 'eth'.

**Comment by karalabe** [comment]
Ah, I also missed that during review.

**Comment by gibtang** [comment]
I am using 1.6.7-stable on my Mac and I got this error
debug.traceBlockByNumber("1666841")
Error: invalid argument 0: hex string without 0x prefix
    at web3.js:3104:20
    at web3.js:6191:15
    at web3.js:5004:36
    at <anonymous>:1:1

debug.traceBlockByNumber(1666841)
Error: invalid argument 0: hex string without 0x prefix
    at web3.js:3104:20
    at web3.js:6191:15
    at web3.js:5004:36
    at <anonymous>:1:1

Seems like i need to convert 16666841 to hex instead. Ok if i modify traceBlockByNumber to accept numbers instead of just hex value?

**Pull request merged**